### DISCUSSION TITLE
Problem with Monaco code Highlighting

### DISCUSSION THREAD

**Discussion by lazkany** [eclipse-langium/langium#1232] [Q&A] [answered]
I have developed a small language with Languim. Everything works okay right now except for syntax highlighting on Monaco.

If I paste my monarch syntax on the Monarch editor online, it works. However, this is not the case when I run it on localhost.

Every other thing like validation, parsing,  works fine. I guess I have a problem with my setup.js that I could not find so far.

My setup.js is below:

Thanks for the great work and efforts.

```ts
import { MonacoEditorLanguageClientWrapper } from './monaco-editor-wrapper/index.js';
import { buildWorkerDefinition } from "./monaco-editor-workers/index.js";
import monarchSyntax from "./syntaxes/syntm.monarch.js";

buildWorkerDefinition('./monaco-editor-workers/workers', new URL('', window.location.href).href, false);

MonacoEditorLanguageClientWrapper.addMonacoStyles('monaco-editor-styles');

const client = new MonacoEditorLanguageClientWrapper();
const editorConfig = client.getEditorConfig();
editorConfig.setMainLanguageId('syntm');

editorConfig.setMonarchTokensProvider(monarchSyntax);

editorConfig.setMainCode(`

SPEC S

In(r0,r1,r2) , Out(g0,g1,g2)

Agent A1 -> In(r0) , Out(g0)
Agent A2 -> In(r1) , Out(g1)
Agent A3 -> In(r2) , Out(g2)

Assume Env /*comments*/
true

Guarantee Sys
G (r2 -> F g2)
`);

editorConfig.theme = 'vs';
editorConfig.useLanguageClient = true;
editorConfig.useWebSocket = false;

const workerURL = new URL('./syntm-server-worker.js', import.meta.url);
console.log(workerURL.href);

const lsWorker = new Worker(workerURL.href, {
    type: 'classic',
    name: 'Syntm Language Server'
});
client.setWorker(lsWorker);

// keep a reference to a promise for when the editor is finished starting, we'll use this to setup the canvas on load
const startingPromise = client.startEditor(document.getElementById("monaco-editor-root"));
```

**Comment by msujew** [comment]
@lazkany Do you get any errors in the console when loading the monarch grammar? We could've accidentally generated an invalid grammar.

**Reply by lazkany** [reply]
No at all, in fact if I paste the syntax and the example here

 https://microsoft.github.io/monaco-editor/monarch.html

I get the expected highlighting.

The syntax is below:

```// Monarch syntax highlighting for the syntm language.
export default {
    keywords: [
        'Agent','Assume','F','G','Guarantee','In','Out','R','SPEC','U','X','false','true'
    ],
    operators: [
        '!','&',',','->','<->','|'
    ],
    symbols: /!|&|\(|\)|,|->|<->|\|/,

    tokenizer: {
        initial: [
            { regex: /[_a-zA-Z][\w_]*/, action: { cases: { '@keywords': {"token":"keyword"}, '@default': {"token":"ID"} }} },
            { include: '@whitespace' },
            { regex: /@symbols/, action: { cases: { '@operators': {"token":"operator"}, '@default': {"token":""} }} },
        ],
        whitespace: [
            { regex: /\s+/, action: {"token":"white"} },
            { regex: /\/\*/, action: {"token":"comment","next":"@comment"} },
            { regex: /\/\/[^\n\r]*/, action: {"token":"comment"} },
        ],
        comment: [
            { regex: /[^/\*]+/, action: {"token":"comment"} },
            { regex: /\*\//, action: {"token":"comment","next":"@pop"} },
            { regex: /[/\*]/, action: {"token":"comment"} },
        ],
    }
};
```

**Reply by msujew** [reply]
Interesting. Can you share a reproducible example? No idea what could be wrong at first glance.

**Reply by cdietrich** [reply]
me neither. looks quite similar to the example i use

**Reply by lazkany** [reply]
I do not know what a reproducible example could be in this case, but I can share my languim grammar.

```
grammar Syntm

entry Model:
    (  
    (name=Module)
    (interfc=Int)*
    (agents+=Agent)*
    (Assumptions=Assumption)*
    (Guarantees=Guarantee)*
    );

Module: 'SPEC' name=ID;
Int:
    'In' '(' sins=In ')'  ','  'Out' '(' souts=Out ')';
In:
    (ins+=Param (',' ins+=Param)*)?;
Out:
    (outs+=Param (',' outs+=Param)*)?;
Agent:
    'Agent'  name=ID  '->' 'In' '(' ains=In ')'  ','  'Out' '(' aouts=Out ')';
Assumption:
    'Assume' name=ID
        (formula+=Phi)+;
Guarantee:
    'Guarantee' name=ID
        (formula+=Phi)+;

Phi: 
    fexp= Simple;
Simple:
    Exp ({infer Phi.left=current} BOP right=Exp)*;
Exp:
     '(' Phi ')' | UPhi | name=ID | name='true' | name='false';

UPhi:
      UOP (op=Exp);
BOP:
    name='&' | name='|' | name='->' | name='<->' | name='U' | name='R';
UOP:
    name='!' | name='X' | name='G' | name='F';
Param:
    name=ID;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

**Reply by cdietrich** [reply]
A repo example would be a git repo that can be cloned to reproduce

**Reply by lazkany** [reply]
Sure, here is a link

https://github.com/SynTEAM-Y/SyntmIDE

Many Thanks

**Reply by cdietrich** [reply]
there is something broken how you copy the js file arround

![Bildschirmfoto 2023-10-19 um 13 18 44](https://github.com/eclipse-langium/langium/assets/238736/4972b72d-7cea-4069-8c09-ff618420b293)

**Reply by lazkany** [reply]
Hi, 

Sorry I am not that experienced. I was only following the tutorial. Okay, I see that this is old, but can't figure out why it did not pick up my new file instead.

**Reply by cdietrich** [reply]
this is the import in your setup.js i assume

**Reply by cdietrich** [reply, marked as answer]
![Bildschirmfoto 2023-10-19 um 13 27 48](https://github.com/eclipse-langium/langium/assets/238736/52b5c997-a1f7-4cca-a0d9-a2a79a0e87da)

**Reply by lazkany** [reply]
<img width="1426" alt="Screenshot 2023-10-19 at 13 43 03" src="https://github.com/eclipse-langium/langium/assets/9892259/dec6aa4a-c086-4d07-95ff-2bc36960f153">

This is how it looks on my laptop!

**Reply by lazkany** [reply]
Okay Thanks that helped a lot. I found the broken thing.

In package.json, there is a command that says: "copy:monarch-syntax": "shx cp -fr ./out/syntaxes ./public/syntaxes",(By the way this is autogenerated)

this creates another syntaxes folder inside and causes the problem.

I changed it to "copy:monarch-syntax": "shx cp -fr ./out/syntaxes ./public/", and it works

Many Thanks

**Reply by cdietrich** [reply]
if this is the case with the latest generator langium i consider this a bug
cc @kaisalmen

**Reply by Yokozuna59** [reply]
It should have been resolved in #1079.
@msujew I guess the [review](https://github.com/eclipse-langium/langium/pull/1079#discussion_r1238609024) didn't resolve the issue.

**Reply by kaisalmen** [reply]
@cdietrich I review and fix this with #1205

**Reply by kaisalmen** [reply]
@msujew @lazkany @cdietrich and @Yokozuna59 this issue is resolved in #1205 already as the copy step is no longer required. The output is directly compiled to the correct directory there.
